src/rebar_plugins.erl

```diff
@@ -172,3 +184,98 @@ validate_plugin(Plugin) ->
+        true ->
+            [];
+        false ->
+            %% Inject source paths for plugins to allow vendoring and umbrella
+            %% top-level declarations
+            BaseDir = rebar_state:dir(State),
+            LibDirs = rebar_dir:project_plugin_dirs(State),
+            Dirs = [filename:join(BaseDir, LibDir) || LibDir <- LibDirs],
+            RebarOpts = rebar_state:opts(State),
+            SrcDirs = rebar_dir:src_dirs(RebarOpts, ["src"]),
+            Found = rebar_app_discover:find_apps(Dirs, SrcDirs, all, State),
+            ?DEBUG("Found local plugins: ~p~n"
+                   "\tusing config: {project_plugin_dirs, ~p}",
+                   [[rebar_utils:to_atom(rebar_app_info:name(F)) || F <- Found],
+                    LibDirs]),
+            PluginsDir = rebar_dir:plugins_dir(State),
+            SetUp = lists:map(fun(App) ->
+                Name = rebar_app_info:name(App),
+                OutDir = filename:join(PluginsDir, Name),
+                prepare_plugin(rebar_app_info:out_dir(App, OutDir))
+            end, Found),
+            rebar_utils:sort_deps(SetUp)
+    end.
+
+is_umbrella(State) ->
+    %% We can't know if this is an umbrella project before running app discovery,
+    %% but plugins are installed before app discovery. So we do a heuristic.
+    %% The lib dirs we search contain things such as apps/, lib/, etc.
+    %% which contain sub-applications. Then there's a final search for the
+    %% local directory ("."), which finds the top-level app in a non-umbrella
+    %% project.
+    %%
+    %% So what we do here is look for the library directories without the ".",
+    %% and if none of these paths exist but one of the src_dirs exist, then
+    %% we know this is not an umbrella application.
+    Root = rebar_dir:root_dir(State),
+    LibPaths = lists:usort(rebar_dir:lib_dirs(State)) -- ["."],
+    SrcPaths = rebar_dir:src_dirs(rebar_state:opts(State), ["src"]),
+    lists:any(fun(Dir) -> [] == filelib:wildcard(filename:join(Root, Dir)) end, LibPaths)
+    andalso
+    lists:all(fun(Dir) -> not filelib:is_dir(filename:join(Root, Dir)) end, SrcPaths).
+
+prepare_plugin(AppInfo) ->
+    %% We need to handle plugins as dependencies to avoid re-building them
+    %% continuously. So here we copy the app directories to the dep location
+    %% and then change the AppInfo record to be redirected to the dep location.
+    AppDir = rebar_app_info:dir(AppInfo),
+    OutDir = rebar_app_info:out_dir(AppInfo),
+    rebar_prv_compile:copy_app_dirs(AppInfo, AppDir, OutDir),
+    Relocated = rebar_app_info:dir(AppInfo, OutDir),
+    case needs_rebuild(AppInfo) of
+        true -> rebar_app_info:valid(Relocated, false);     % force recompilation
+        false -> rebar_app_info:valid(Relocated, undefined) % force revalidation
+    end.
+
+top_level_deps(Apps) ->
+    RawDeps = lists:foldl(fun(App, Acc) ->
+        %% Only support the profiles we would with regular plugins?
```

Comment:
The `?` makes it sound like this does support profiles besides default and prod? But the code looks like it is filtering by those 2 profiles?

Preferably we'd support all profiles since this is potentially used not for vendoring but for normal development. Like if I wanted a plugin to help with development on OpenTelemetry which is an umbrella.

Reply:
Yeah, I was expecting that default plugins would only fetch the plugins in default+prod mode, like any dependency. The top-level plugins are already selected by the time we run this here, this is about hoisting the plugins' own deps and plugins into the project as well. I'm unsure whether we need to handle more profiles than that, because I'm not sure we want to get dependencies' test plugins into a regular space.

Reply:
aaah, the plugins deps, ok.

Reply:
Tho if I'm developing a project and want a plugin in the umbrella I'd expect its tests to run if I ran `rebar3 ct` at the top level if the plugin is under `apps`, as opposed to another directory like `plugins` where I would expect a separation of concerns.

Reply:
Yeah that's fair. I'll have to look at bringing in the profile expanding of deps for plugins I guess.

Reply:
Take a look again, made it fetch more profiles.